commit ee9e01bf527be3c3916679b65e86ae84ffff47a0
Author: Isaac Davis <isaac.davis@joyent.com>
Date:   2019-07-16T20:37:55+00:00 (3 months ago)
    
    MANTA-4400 Add registrar config to boray
    Reviewed by: Kelly McLaughlin <kelly.mclaughlin@joyent.com>
    Approved by: Kelly McLaughlin <kelly.mclaughlin@joyent.com>

diff --git a/Makefile b/Makefile
index 9b430ce..0739f00 100644
--- a/Makefile
+++ b/Makefile
@@ -70,6 +70,20 @@ release: all deps docs $(SMF_MANIFESTS)
 	cd $(RELSTAGEDIR) && $(TAR) -I pigz -cf $(ROOT)/$(RELEASE_TARBALL) root site
 	@rm -rf $(RELSTAGEDIR)
 
+#
+# We include a pre-substituted copy of the template in our built image so that
+# registrar doesn't go into maintenance on first boot. Then we ship both the
+# .in file and this "bootstrap" substituted version. The boot/setup.sh script
+# will perform this replacement again during the first boot, replacing @@PORTS@@
+# with the real port.
+#
+# This default value should be kept in sync with the default value in
+# boot/setup.sh and sapi_manifests/boray/template
+#
+#
+sapi_manifests/registrar/template: sapi_manifests/registrar/template.in
+	sed -e 's/@@PORTS@@/2030/g' $< > $@
+
 .PHONY: publish
 publish: release
 	mkdir -p $(ENGBLD_BITS_DIR)/$(NAME)
diff --git a/boot/setup.sh b/boot/setup.sh
index 5b55dff..cbab149 100755
--- a/boot/setup.sh
+++ b/boot/setup.sh
@@ -22,14 +22,79 @@ DIR="$( cd -P "$( dirname "$SOURCE" )" && pwd )"
 PROFILE=/root/.bashrc
 PG_USER=boray
 SVC_ROOT=/opt/smartdc/boray
-ZONE_UUID=`/usr/bin/zonename`
+ZONE_UUID=$(/usr/bin/zonename)
+SAPI_CONFIG=
+SAPI_URL=$(mdata-get SAPI_URL)
+[[ -n $SAPI_URL ]] || fatal "no SAPI_URL found"
+#
+# Wait 10 seconds for dns to become operational. This is displeasing, but
+# otherwise we'll encounter dns resolution errors. This logic is lifted from
+# electric-moray.
+#
+sleep 10
 
 export PATH=$SVC_ROOT/bin:$SVC_ROOT/build/node/bin:/opt/local/bin:/usr/sbin/:/usr/bin:$PATH
 
-function manta_setup_boray {
+function get_sapi_config {
+    local sapi_res
+    local err
+    #
+    # Load the zone's full config object from SAPI. If this request fails, it
+    # will be retried until the smf(5) method timeout expires for the
+    # "mdata:execute" service.
+    #
+    # We do this because we need to access SAPI config variables in this script,
+    # and it's more difficult to get them from the boray config toml file.
+    #
+    while :; do
+        if ! sapi_res=$(curl --max-time 60 --ipv4 -sSf \
+          -H 'Accept: application/json' -H 'Content-Type: application/json' \
+          "${SAPI_URL}/configs/${ZONE_UUID}"); then
+            printf 'WARNING: could not download SAPI config (retrying)\n' >&2
+            sleep 2
+            continue
+        fi
+
+        SAPI_CONFIG=$(json -H <<< "${sapi_res}")
+
+        err=$(json 'code' <<< "${SAPI_CONFIG}")
+        if [[ -n "${err}" ]]; then
+            printf 'WARNING: error parsing SAPI config (%s) (retrying)\n' >&2
+            sleep 2
+            continue
+        fi
+        break
+    done
+}
+
+function setup_boray {
+    local port
+    local RTPL
+
+    #
+    # The default port value here must be kept in sync with the default value of
+    # BORAY_SERVER_PORT in sapi_manifests/boray/template and the Makefile.
+    #
+    port=$(json metadata.BORAY_SERVER_PORT <<< "${SAPI_CONFIG}")
+    [[ -n "${port}" ]] || port='2030'
+
+    #
+    # Regenerate the registrar config with the real port included
+    # (the bootstrap one just includes 2030, the default value)
+    #
+    RTPL=$SVC_ROOT/sapi_manifests/registrar/template
+    sed -e "s/@@PORTS@@/${port}/g" ${RTPL}.in > ${RTPL}
+
+    #
+    # Wait until config-agent updates registrar's config before restarting
+    # registrar.
+    #
+    svcadm disable -s config-agent
+    svcadm enable -s config-agent
+    svcadm restart registrar
+
     svccfg import /opt/smartdc/boray/smf/manifests/boray.xml
-    svcadm enable boray
-    [[ $? -eq 0 ]] || fatal "unable to start boray"
+    svcadm enable boray || fatal "unable to start boray"
 }
 
 source ${DIR}/scripts/util.sh
@@ -46,7 +111,8 @@ manta_common_setup "boray" 0
 manta_ensure_zk
 
 echo "Setting up Boray"
-manta_setup_boray
+get_sapi_config
+setup_boray
 
 manta_common_setup_end
 
diff --git a/sapi_manifests/boray/template b/sapi_manifests/boray/template
index eb3b6b7..61a3a46 100644
--- a/sapi_manifests/boray/template
+++ b/sapi_manifests/boray/template
@@ -17,6 +17,10 @@ host = "0.0.0.0"
 port = {{BORAY_SERVER_PORT}}
 {{/BORAY_SERVER_PORT}}
 {{^BORAY_SERVER_PORT}}
+#
+# This default value should be kept in sync with the default value in
+# boot/setup.sh and the Makefile.
+#
 port = 2030
 {{/BORAY_SERVER_PORT}}
 
diff --git a/sapi_manifests/registrar/manifest.json b/sapi_manifests/registrar/manifest.json
new file mode 100644
index 0000000..f57ee4f
--- /dev/null
+++ b/sapi_manifests/registrar/manifest.json
@@ -0,0 +1,5 @@
+{
+    "name": "registrar",
+    "path": "/opt/smartdc/registrar/etc/config.json",
+    "master": true
+}
\ No newline at end of file
diff --git a/sapi_manifests/registrar/template.in b/sapi_manifests/registrar/template.in
new file mode 100644
index 0000000..5c548c0
--- /dev/null
+++ b/sapi_manifests/registrar/template.in
@@ -0,0 +1,31 @@
+{
+  "registration": {
+    "domain": "{{SERVICE_NAME}}",
+    "type": "moray_host",
+    "service": {
+      "type": "service",
+      "service": {
+        "srvce": "_boray",
+        "proto": "_tcp",
+        "port": 2020
+      },
+      "ttl": 60
+    },
+    "ttl": 60,
+    "ports": [@@PORTS@@]
+  },
+
+  "zookeeper": {
+    "servers": [
+      {{#ZK_SERVERS}}
+        {
+          "host": "{{host}}",
+          "port": {{port}}
+        }{{^last}}, {{/last}}
+      {{/ZK_SERVERS}}
+    ],
+    "timeout": 60000
+  },
+
+  "adminIp": "{{auto.MANTA_IP}}"
+}
